### PR TITLE
gh-118934: Make PyEval_GetLocals return borrowed reference

### PR DESCRIPTION
**Pull request by gaogaotiantian**
Per discussion with @markshannon in Pittsburgh, we simulate what we did pre-PEP667. We store the locals dict in `frame->f_locals` so that we can return a borrowed reference. The `frame->f_locals` will be cleared when the frame is released. This way we can keep the behavior of `PyEval_GetLocals()`.

* Issue: gh-118934

### CODE REVIEW

**Comment by gaogaotiantian** [comment]
Just realized this will overwrite `f_locals` when it's called inside a comprehension of a non-optimized scope. We might have to use a new member in FrameObject to store this.

File: Python/ceval.c
Comment (colesbury) [inline Python/ceval.c]: formatting: place `else if` on a new line after `}`

File: Python/ceval.c
Comment (colesbury) [inline Python/ceval.c]: formatting: `PyObject *ret`

File: Python/ceval.c
Comment (colesbury) [inline Python/ceval.c:2563]: Is this always a `dict` (not just a mapping)? The docs say that this function returns a dictionary:

https://docs.python.org/3/c-api/reflection.html#c.PyEval_GetLocals

**Review by colesbury — COMMENTED** [review]
A few formatting suggestions and a question about the return value

File: Python/ceval.c
Comment (gaogaotiantian) [reply Python/ceval.c:2563]: Unfortunately no. This is probably a neglection in the docs. `exec()` takes [any mapping object](https://docs.python.org/3/library/functions.html#exec) as `locals` and inside that scope, `locals()` will return whatever mapping object that is passed in. We also can't convert it to dict because that would be another disaster. Maybe we should fix the docs? However, in normal cases, `locals()` (and `PyEval_GetLocals`) returns a dict.

However, I need to change more than this. I had to add an extra field in `FrameObject` because of the matter I mentioned above, so this is not the final version :(

Comment (ncoghlan) [reply Python/ceval.c:2563]: Yeah, I just partially fixed the docs around this in https://github.com/python/cpython/pull/119893 so they're at least correct for optimized scopes (well, they will be once this PR lands).

As @gaogaotiantian noted, there are still some mentions of `dictionary` in relation to `locals` that aren't strictly true, though (since class scopes aren't always dictionaries)

File: Python/ceval.c
Comment (ncoghlan) [inline Python/ceval.c]: We don't actually want to make the snapshot in this case - we want to cache and return the proxy as described in PEP 667. It's the replacement for `PyFrame_LocalsToFast()` no longer doing anything.

Comment (gaogaotiantian) [reply Python/ceval.c]: `PyEval_GetLocals()` returns a dict/mapping as `locals()`. Maybe you were thinking of `PyFrame_GetLocals()`?

Comment (ncoghlan) [reply Python/ceval.c]: Turns out there is ambiguity on this point in PEP 667 (two references to caching a proxy instance, one to caching a snapshot): https://github.com/python/cpython/issues/118934#issuecomment-2143330127

File: Python/ceval.c
Comment (ncoghlan) [reply Python/ceval.c:2563]: Also see https://github.com/python/cpython/issues/118934#issuecomment-2143330127 in relation to the docs and the ambiguity over the expected behaviour of `PyEval_GetLocals` in optimized scopes.

File: Python/ceval.c
Comment (ncoghlan) [inline Python/ceval.c]: Regardless of whether it's a proxy or a shared snapshot, the new `f_locals_cache` field needs to be checked to see if it is NULL before getting populated again.

Comment (gaogaotiantian) [reply Python/ceval.c]: Sorry which variable do you refer to?

File: Python/ceval.c
Comment (ncoghlan) [reply Python/ceval.c:2563]: https://github.com/python/cpython/pull/119932 fixed up the lingering mentions of locals always being a dict (as well as aligning the `PyEval_GetLocals` docs with the way things will work once this PR is complete & merged).

File: Python/ceval.c
Comment (ncoghlan) [reply Python/ceval.c]: https://github.com/python/cpython/pull/119932 aligned the `PyEval_GetLocals` docs with the way it will work once this PR is completed and merged.

Comment (ncoghlan) [reply Python/ceval.c]: `PyObject *ret = PyDict_New()`

When `f_locals_cache` is already set, it should be reused here rather than allocating a new dictionary (otherwise the desired reference borrowing semantics won't be provided, as subsequent `PyEval_GetLocals()` calls on the same frame would invalidate the reference pointers returned by previous calls)

File: Python/ceval.c
Comment (ncoghlan) [inline Python/ceval.c]: Edit: removed the "store in a list" suggestion, since the SC opted to keep the Python 3.12 semantics for this API.

File: Python/ceval.c
Comment (ncoghlan) [reply Python/ceval.c]: Code suggestion added below.

Comment (gaogaotiantian) [reply Python/ceval.c]: Hmm, I don't agree here. I don't think it's our job to make sure the borrowed reference is "always valid", that's why we need new reference. The user should use borrowed reference in their own risk. For all of our C APIs that return a borrowed reference, the user can do something to invalidate the reference, that's just how borrowed references work.

I think the goal here is to make `PyEval_GetLocals()` mimic `locals()` or `PyEval_GetFrameLocals()`, with the only difference of returned reference type. And to avoid confusion, it should mimic `locals()` in 3.13, not 3.12 (or `PyEval_GetLocals()` in 3.12). We changed the behavior of `locals()` in 3.13 so that it returns a snapshot (that will not be updated) of the current frame locals, and I think the proper behavior of `PyEval_GetLocals()` should be the same - the dict it get should not be updated either, otherwise it will just confuse users.

Comment (gaogaotiantian) [reply Python/ceval.c]: And I really think we should confirm the behavior before updating any documentation. I don't believe making `PyEval_GetLocals()` an exception that is different than `locals()` is the way to go. It should be as simple as - it performs exactly like `locals()` and `PyEval_GetFrameLocals()`, except for the returned reference type. And the original goal is to deprecate it, along with `PyEval_GetGlobals()` and `PyEval_GetBuiltins()`.

Comment (gaogaotiantian) [reply Python/ceval.c]: I know there's a rough implementation in PEP 667, but that implementation does not update the locals cache at all so maybe that's not the bahavior we wanted. If we can't agree on the bahavior, we could involve more people discussing about this. @markshannon maybe as the original author?

Comment (ncoghlan) [reply Python/ceval.c]: The main thing I want to ensure is that the following sequence of calls doesn't pose a risk of segfaulting the interpreter:

```
    PyObject *ref1 = PyEval_GetLocals();
    PyObject *ref2 = PyEval_GetLocals();  // This would likely be in a called function rather than inline
    // Do something with ref1 (that should be safe, since the frame object the reference is borrowed from is still alive)
```

However, I realised there's a reasonably straightforward way to satisfy all of the following objectives:

* make every call to `PyEval_GetLocals` return a new snapshot on optimised frames
* return a borrowed reference
* keep the borrowed references valid until the frame object goes away

Rather than caching the returned snapshot directly, the snapshots can instead be cached inside a list object. Multiple calls just append new entries to the end of the list. Still thoroughly deprecation worthy, but the segfault risk is gone, while still updating the semantics to match the 3.13 `locals()` behaviour.

I've amended the code suggestion below to implement that behaviour.

(We could optionally delay creating the intermediate list until the second call to `PyEval_GetLocals()` on the same frame, but I don't think the memory saving would be big enough to be worth the extra code complexity in a deprecated API).

(as far as the docs go, even though they're going to need further updates, what they say now is closer to reality than what they said after I updated them based on the PEP text that said `PyEval_GetLocals` would start returning a write-through proxy instance. I won't tinker further until this is resolved, though)

File: Python/ceval.c
Comment (ncoghlan) [reply Python/ceval.c]: ~~Suggestion updated to store returned snapshots in a list, so the function can switch to the independent snapshot semantics of `locals()` while still returning a borrowed reference that remains valid for the life of the frame object.~~ (Edit: this option was removed as per SC feedback)

File: Python/ceval.c
Comment (gaogaotiantian) [reply Python/ceval.c]: This was my original thought, but I don't favor that idea now after I talked with Mark (and I still think we should get his input).

I think the key argument is, do we need to ensure this:
```C
    PyObject *ref1 = PyEval_GetLocals();
    PyObject *ref2 = PyEval_GetLocals();  // This would likely be in a called function rather than inline
    // Do something with ref1 (that should be safe, since the frame object the reference is borrowed from is still alive)
```

Like I said, it's trivial for user to crash CPython at C level. The user is responsible to make sure the "borrowed reference" is alive. If the user wants to make sure, the should `Py_INCREF(ref1)`, that's the only reasonable thing to do.

For example,
```C
PyObject* item = PyDict_GetItem(dct, key);
PyDict_SetItem(dct, key, new_value);
// Should we make sure item is alive here?
```

Notice in the example above, the `dct` object is not freed, just like the `FrameObject` in the `PyEval_GetLocals()` case. There's no obligation for us to make sure all the objects "calculated" from object A should live until object A is released.

borrowed reference could be invalidated in many scenarios and users should proceed with precaution. I think that's why we want to replace our borrowed reference APIs with new reference. I don't think that we should do the list here just to prevent users from crashing themselves with dangerous code.

The cost is not huge, I agree, but the memory will be used until the frame is released (if user uses `PyEval_GetLocals()` in a loop inside a function? That's `O(N)` increment for memory when the user thinks they are using borrowed reference).

That's my thoughts about this issue. Again I might not be correct, and I'm open to other suggestions and opinions.

Comment (ncoghlan) [reply Python/ceval.c]: The thing with borrowed references is that they're borrowed *from* another object, and in the case of `PyEval_GetLocals()` that object is the frame object. In previous versions, as long as the frame object was still alive, then the borrowed locals reference remained valid, including in the presence of multiple calls to `PyEval_GetLocals()`. (Mutating a container isn't the same situation, since it's known that mutation operations on containers can invalidate borrowed references to items within the container, while refreshing a cache shouldn't risk making references to the cache itself invalid)

If we don't preserve that property, then C code that worked fine on Python 3.12 will likely start segfaulting on Python 3.13. If PEP 667 had said that was going to happen, then I expect the SC would have made a request to figure out a way to avoid the segfault risk. (Neither the text nor the example code in PEP 667 carried a segfault risk, they just differed in whether the cached value was a proxy creating a reference cycle or a snapshot dict that no other APIs would access)

Don't get me wrong, I don't think using a list of references is a *good* option, I just think it's the *least bad* one if we want to convert the function's semantics over to those of the Python 3.13 `locals` builtin.

I'm also fine with leaving it back on the Python 3.12 `locals` builtin semantics, where repeated calls refresh a shared snapshot (I actually prefer that option, since existing users of the API are necessarily familiar with that behaviour, it doesn't increase memory usage relative to Python 3.12, and new users should be using `PyEval_GetFrameLocals()` anyway, so if the weird legacy semantics push them away from `PyEval_GetLocals()` that's a good thing). 

Of all the APIs deprecated in PEP 667, `PyEval_GetLocals()` is the only one I'd like to see upgraded to a hard deprecation such that it eventually goes away (while the SC already said "no" to a hard deprecation in 3.13, we'll hopefully be able to reconsider that in 3.14).

File: Python/ceval.c
Comment (ncoghlan) [reply Python/ceval.c]: SC feedback is that we should keep the Python 3.12 `locals` semantics for this one API (since existing users are familiar with it anyway, and it works better with the borrowed reference than the Python 3.13 semantics do):

```suggestion
        PyObject *ret = f->f_locals_cache;
        if (ret == NULL) {
            PyObject *ret = PyDict_New();
            if (ret == NULL) {
                Py_DECREF(locals);
                return NULL;
            }
            f->f_locals_cache = ret;
        }
        if (PyDict_Update(ret, locals) < 0) {
            // At this point, if the cache dict is broken, it will stay broken, as
            // trying to clean it up or replace it will just cause other problems
            ret = NULL;
        }
        Py_DECREF(locals);
        return ret;
```

File: Python/ceval.c
Comment (gaogaotiantian) [reply Python/ceval.c]: I understand your point. My argument is that it would be a very rare case where the user updates the cache with `PyEval_GetLocals()` and still try to access the previous one. The list solution is so bad (and it's not your fault it's just what we have to do to keep `locals()` behavior + save all possible segfaults) that I would also lean to keeping the 3.12 behavior just to make the code cleaner.

I still think we need more eyes on this to reach a conclusion. Our possible solutions are all very clear and we both know the pros and cons for each, but we have different opinions and evaluations. Maybe involving more people, even SC in this decision, could solve the problem for good. I won't be persistent if most of the people think the alternatives are better.

Comment (ncoghlan) [reply Python/ceval.c]: I don't think anyone would be directly calling PyEval_GetLocals multiple times in the same function, but I *do* consider it entirely plausible that they may be calling other APIs in their application or extension module that also call PyEval_GetLocals independently.

So I think our disagreement boils down to two points:

* I think the above scenario, while genuinely unlikely, is plausible enough that we shouldn't risk making it segfault. You consider it so unlikely that it really isn't worth worrying about.
* You see more value in making PyEval_GetLocals consistent with the semantics of the rest of Python 3.13 than I do

I'll summarise this on the open SC ticket, and mention that ticket in the release discussion channel in Discord. 

(Originally I thought caching and returning a proxy instance offered the best of both worlds, but changed my mind when I realised it posed a different set of segfault risks if anyone had assumed it was a dict without properly checking the runtime type)

Comment (ncoghlan) [reply Python/ceval.c]: For the SC question, I'll present all three options:

* keep Python 3.12 semantics
* adopt Python 3.13 semantics by invalidating previously borrowed references on repeated calls
* adopt Python 3.13 semantics by keeping a list of references

and let them know neither of us actually prefer the third option, but it does exist

Comment (gaogaotiantian) [reply Python/ceval.c]: That is very thorough and represents exactly what I thought. Thank you for putting together the options (I'm not very good with the words and the documentation LOL)

Comment (ncoghlan) [reply Python/ceval.c]: Your explanations here have been entirely clear - I'm definitely at a point where I think any of the three options would actually be mostly OK (the only adjustment I'd argue for is that if we did go down the `Py_XSETREF` path we add a hard deprecation warning that gets triggered when `PyEval_GetLocals()` is called on an optimised frame with `f_locals_cache` already set. That way if folks did manage to get a segfault, they'd hopefully see the deprecation warning first)

Summary has been posted here (trying to be as fair to both points of view as I can): https://github.com/python/steering-council/issues/245#issuecomment-2148820321

I'll also ping @Yhg1s as release manager in the Discord release-discussions channel.

Comment (ncoghlan) [reply Python/ceval.c]: @markshannon recommended in https://github.com/python/steering-council/issues/245#issuecomment-2155093992, that we trust the original PEP text over its sample code and cache a locals proxy instance instead of a snapshot (i.e. bring `PyEval_GetLocals()` down on the side of aligning with `f.f_locals` access rather than with `locals()`)

I'm certainly fine with that option (going back to those docs mostly involves restoring text I had previously revised), so if @gaogaotiantian is also OK with it, there's no longer a disagreement for the SC to resolve.

Comment (ncoghlan) [reply Python/ceval.c]: SC has confirmed that `PyEval_GetLocals` should retain its 3.12 behaviour: https://github.com/python/steering-council/issues/245#issuecomment-2179005461

**Comment by ncoghlan** [comment]
@gaogaotiantian I merged the suggestion that reverts back to the Python 3.12 semantics, so we can get a CI run on that version of the refleak fix.

**Review by ncoghlan — APPROVED** [review]
@gaogaotiantian CI is happy with going back to the Python 3.12 behaviour, so if you're happy I think this is good to merge for the final beta.

**Pull request merged**
